### PR TITLE
Implement BatchMatMul with Numpy-style batch broadcast semantics

### PR DESCRIPTION
**Pull request by jamesr66a**
ONNX has decided to implement a single MatMul operator that borrows semantics from np.matmul: https://docs.scipy.org/doc/numpy-1.13.0/reference/generated/numpy.matmul.html

This PR introduces a new op that we can target for ONNX that mimics the numpy-style broadcast semantics

### CODE REVIEW

File: caffe2/operators/numpy_batch_matmul_op.h [hunk 1]
```diff
@@ -0,0 +1,240 @@
...
+        dims_B.push_back(1);
+        ndims_B = 2;
+        B_broadcasted = true;
+      }
+      // matrix-matrix with batches
+      // [B1..., M, K] * [B2..., K, N] -> [B..., M, N]
+      // In the event that A or B are one-dimensional, the trailing or leading
+      // 1 is not added to the output tensor's size.
+
+      // First step: partition the tensors into inner and outer blocks.
```
Comment (dzhulgakov) [inline caffe2/operators/numpy_batch_matmul_op.h]: it's actually not a complete numpy semantics (it also supports '1' dimensions). But agree, let's skip it for now

File: caffe2/operators/numpy_batch_matmul_op.h [hunk 2]
```diff
@@ -0,0 +1,240 @@
...
+      // flags
+      size_t M, N, K;
+      if (trans_a_) {
+        M = dims_A[ndims_A - 1];
+        K = dims_A[ndims_A - 2];
+      } else {
+        M = dims_A[ndims_A - 2];
+        K = dims_A[ndims_A - 1];
+      }
+      if (trans_b_) {
```
Comment (dzhulgakov) [inline caffe2/operators/numpy_batch_matmul_op.h]: do you have a check somewhere that another dimension of dims_B matches?

File: caffe2/operators/numpy_batch_matmul_op.h [hunk 3]
```diff
@@ -0,0 +1,240 @@
...
+      } else {
+        N = dims_B[ndims_B - 1];
+      }
+
+      // Calculate output tensor shapes [B..., (M), (N)]
+      // Batch dimensions will be broadcasted out to those of the longer tensor
+      // A or B. Either M or N are optional if A or B, respectively are 1-D.
+      std::vector<TIndex> new_dims;
+      if (ndims_A >= ndims_B) {
+        std::copy(
```
Comment (dzhulgakov) [inline caffe2/operators/numpy_batch_matmul_op.h]: nit: or new_dims.assign(dims_A.begin(), dims_A.end() - 2)

File: caffe2/operators/numpy_batch_matmul_op.h [hunk 4]
```diff
@@ -0,0 +1,240 @@
...
+            dims_A.begin(), dims_A.end() - 2, std::back_inserter(new_dims));
+      } else {
+        std::copy(
+            dims_B.begin(), dims_B.end() - 2, std::back_inserter(new_dims));
+      }
+      if (!A_broadcasted)
```
Comment (dzhulgakov) [inline caffe2/operators/numpy_batch_matmul_op.h]: nit: {}

**Review by dzhulgakov — COMMENTED** [review]
I think this op can be a drop-in replacement for the current BatchMatMul that hard-codes 3-D tensors and doesn't do broadcasting, am I wrong?

File: caffe2/operators/numpy_batch_matmul_op.h (hunk 1 above)
Comment (jamesr66a) [reply caffe2/operators/numpy_batch_matmul_op.h]: Could you elaborate on this? Not sure what you're referring to

**Pull request closed without merging**